### PR TITLE
feat: Add MD5 Hashing Algorithm

### PR DESCRIPTION
**Pull request by tGautot**
#### Description of Change
Added my own implementation of the MD5 Hashing Algorithm/protocol to the hashing directory

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: Thanks for giving me a reason to finally implement it in cpp

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1595"></a>

### CODE REVIEW

File: hashing/md5.cpp [hunk 1]
```diff
@@ -0,0 +1,352 @@
+/**
+ * @file md5.cpp
```
Comment (Panquesito7) [inline hashing/md5.cpp:2]:
```suggestion
 * @file
```

File: hashing/md5.cpp [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+ * @author tGautot (https://github.com/tGautot)
```
Comment (Panquesito7) [inline hashing/md5.cpp:3]:
```suggestion
 * @author [tGautot](https://github.com/tGautot)
```

File: hashing/md5.cpp [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+ * @brief Simple C++ implementation of the [MD5 Hashing
+ * Algorithm](https://en.wikipedia.org/wiki/MD5)
+ *
+ *
```
Comment (Panquesito7) [inline hashing/md5.cpp:7]:
```suggestion
```

File: hashing/md5.cpp [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+ * MD5 is one of the most used hashing algorithm there is. Some of its
+ * use cases are:
+ *  1. Providing checksum for downloaded software
+ *  2. Store salted password
+ *
+ * However MD5 has be know to be cryptographically weak for quite some
+ * time, yet it is still widely used. This weakness was exploited by the
+ * [Flame Malware](https://en.wikipedia.org/wiki/Flame_(malware)) in 2012
+ *
+ * ### The Algorithm
```
Comment (Panquesito7) [inline hashing/md5.cpp:23]:
```suggestion
 * ### Algorithm
```

File: hashing/md5.cpp [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+#include <array>  // To avoid c-style array (thanks clang-tidy)
+#include <cassert>
+#include <cstring>  // Used for std::memcopy
+#include <iostream>
+#include <string>
+#include <vector>  // To avoid c-style array of runtime size (thanks clang-tidy)
+
+/**
+ * @namespace MD5
+ *
```
Comment (Panquesito7) [inline hashing/md5.cpp:55]:
```suggestion
```

File: hashing/md5.cpp [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+ * @brief MD5 algorithm for hashing plus some function it uses
```
Comment (Panquesito7) [inline hashing/md5.cpp:56]:
```suggestion
 * @brief Functions for the [MD5](https://en.wikipedia.org/wiki/MD5) algorithm implementation
```

File: hashing/md5.cpp [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+ *
+ */
+namespace md5 {
+
```
Comment (Panquesito7) [inline hashing/md5.cpp:60]:
```suggestion
```

File: hashing/md5.cpp [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+#include <algorithm>
+#include <array>  // To avoid c-style array (thanks clang-tidy)
+#include <cassert>
+#include <cstring>  // Used for std::memcopy
+#include <iostream>
+#include <string>
+#include <vector>  // To avoid c-style array of runtime size (thanks clang-tidy)
+
+/**
```
Comment (Panquesito7) [inline hashing/md5.cpp:53]:
```suggestion
/**
 * @namespace hashing
 * @brief Hashing algorithms
 */
namespace hashing {
/**
```

File: hashing/md5.cpp [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+}
+
+/**
+ * @brief Checks whether integers are stored as big endian or not
+ *
+ * @note Taken from [this](https://stackoverflow.com/a/1001373) stackoverflow
+ * post
+ *
+ * @return true
+ * @return false
```
Comment (Panquesito7) [inline hashing/md5.cpp:79]: Returns true IF...? Returns false IF...? Please write that here.

File: hashing/md5.cpp [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+ * @return uint32_t The rotated integer
+ */
+uint32_t leftRotate32bits(uint32_t n, std::size_t rotate) {
+    return (n << rotate) | (n >> (32 - rotate));
+}
+
+/**
+ * @brief Checks whether integers are stored as big endian or not
+ *
+ * @note Taken from [this](https://stackoverflow.com/a/1001373) stackoverflow
```
Comment (Panquesito7) [inline hashing/md5.cpp:75]:
```suggestion
 * @note Taken from [this](https://stackoverflow.com/a/1001373) StackOverflow
```

File: hashing/md5.cpp [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+ * @param rotate How many bits for the rotation
+ * @return uint32_t The rotated integer
+ */
+uint32_t leftRotate32bits(uint32_t n, std::size_t rotate) {
+    return (n << rotate) | (n >> (32 - rotate));
+}
+
+/**
+ * @brief Checks whether integers are stored as big endian or not
+ *
```
Comment (Panquesito7) [inline hashing/md5.cpp:74]:
```suggestion
```

File: hashing/md5.cpp [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+        uint32_t i;
+        std::array<char, 4> c;
+    } bint = {0x01020304};
+
+    return bint.c[0] == 1;
+}
+
+/**
+ * @brief Sets 32-bit integer to little-endian if needed
+ *
```
Comment (Panquesito7) [inline hashing/md5.cpp:92]:
```suggestion
```

File: hashing/md5.cpp [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+ * @param n
+ * @return uint32_t
```
Comment (Panquesito7) [inline hashing/md5.cpp:94]: Missing documentation.

File: hashing/md5.cpp [hunk 14]
```diff
@@ -0,0 +1,352 @@
...
+    }
+    // Machine works on little endian, no need to change anything
+    return n;
+}
+
+/**
+ * @brief Sets 64-bit integer to little-endian if needed
+ *
+ * @param n
+ * @return uint32_t
```
Comment (Panquesito7) [inline hashing/md5.cpp:109]: Missing documentation.

File: hashing/md5.cpp [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+/**
+ * @brief Converts the string to bytestring and calls the main algorithm
+ *
+ * @param message Plain character message to hash
+ * @return void* Pointer to the MD5 signature
+ */
+void* hash(const std::string& message) {
+    return hash_bs(&message[0], message.size());
+}
+}  // namespace md5
```
Comment (Panquesito7) [inline hashing/md5.cpp:306]:
```suggestion
}  // namespace md5
}  // namespace hashing
```

File: hashing/md5.cpp [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+
+void test() {
```
Comment (Panquesito7) [inline hashing/md5.cpp:308]:
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
static void test() {
```

File: hashing/md5.cpp [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+    void* sig = md5::hash("");
```
Comment (Panquesito7) [inline hashing/md5.cpp:309]:
```suggestion
    void* sig = hashing::md5::hash("");
```

File: hashing/md5.cpp [hunk 18]
```diff
@@ -0,0 +1,352 @@
...
+    std::cout << "Hashing empty string" << std::endl;
+    std::cout << md5::sig2hex(sig) << std::endl << std::endl;
```
Comment (Panquesito7) [inline hashing/md5.cpp:311]:
```suggestion
    std::cout << hashing::md5::sig2hex(sig) << std::endl << std::endl;
```

File: hashing/md5.cpp [hunk 19]
```diff
@@ -0,0 +1,352 @@
...
+    assert(md5::sig2hex(sig).compare("d41d8cd98f00b204e9800998ecf8427e") == 0);
```
Comment (Panquesito7) [inline hashing/md5.cpp:312]:
```suggestion
    assert(hashing::md5::sig2hex(sig).compare("d41d8cd98f00b204e9800998ecf8427e") == 0);
```

File: hashing/md5.cpp [hunk 20]
```diff
@@ -0,0 +1,352 @@
...
+
+    void* sig2 = md5::hash("The quick brown fox jumps over the lazy dog");
```
Comment (Panquesito7) [inline hashing/md5.cpp:314]:
```suggestion
    void* sig2 = hashing::md5::hash("The quick brown fox jumps over the lazy dog");
```

File: hashing/md5.cpp [hunk 21]
```diff
@@ -0,0 +1,352 @@
...
+    std::cout << "Hashing The quick brown fox jumps over the lazy dog"
+              << std::endl;
+    std::cout << md5::sig2hex(sig2) << std::endl << std::endl;
```
Comment (Panquesito7) [inline hashing/md5.cpp:317]:
```suggestion
    std::cout << hashing::md5::sig2hex(sig2) << std::endl << std::endl;
```

File: hashing/md5.cpp [hunk 22]
```diff
@@ -0,0 +1,352 @@
...
+    assert(md5::sig2hex(sig2).compare("9e107d9d372bb6826bd81d3542a419d6") == 0);
```
Comment (Panquesito7) [inline hashing/md5.cpp:318]:
```suggestion
    assert(hashing::md5::sig2hex(sig2).compare("9e107d9d372bb6826bd81d3542a419d6") == 0);
```

File: hashing/md5.cpp [hunk 23]
```diff
@@ -0,0 +1,352 @@
...
+
+    void* sig3 = md5::hash("The quick brown fox jumps over the lazy dog.");
+    std::cout << "Hashing "
+                 "The quick brown fox jumps over the lazy dog."
+              << std::endl;
+    std::cout << md5::sig2hex(sig3) << std::endl << std::endl;
```
Comment (Panquesito7) [inline hashing/md5.cpp:324]:
```suggestion
    std::cout << hashing::md5::sig2hex(sig3) << std::endl << std::endl;
```

File: hashing/md5.cpp [hunk 24]
```diff
@@ -0,0 +1,352 @@
...
+    assert(md5::sig2hex(sig3).compare("e4d909c290d0fb1ca068ffaddf22cbd0") == 0);
```
Comment (Panquesito7) [inline hashing/md5.cpp:325]:
```suggestion
    assert(hashing::md5::sig2hex(sig3).compare("e4d909c290d0fb1ca068ffaddf22cbd0") == 0);
```

File: hashing/md5.cpp [hunk 25]
```diff
@@ -0,0 +1,352 @@
...
+    std::cout << md5::sig2hex(sig) << std::endl << std::endl;
+    assert(md5::sig2hex(sig).compare("d41d8cd98f00b204e9800998ecf8427e") == 0);
+
+    void* sig2 = md5::hash("The quick brown fox jumps over the lazy dog");
+    std::cout << "Hashing The quick brown fox jumps over the lazy dog"
+              << std::endl;
+    std::cout << md5::sig2hex(sig2) << std::endl << std::endl;
+    assert(md5::sig2hex(sig2).compare("9e107d9d372bb6826bd81d3542a419d6") == 0);
+
+    void* sig3 = md5::hash("The quick brown fox jumps over the lazy dog.");
```
Comment (Panquesito7) [inline hashing/md5.cpp:320]:
```suggestion
    void* sig3 = hashing::md5::hash("The quick brown fox jumps over the lazy dog.");
```

File: hashing/md5.cpp [hunk 26]
```diff
@@ -0,0 +1,352 @@
...
+    std::cout << "Hashing "
+                 "The quick brown fox jumps over the lazy dog."
+              << std::endl;
+    std::cout << md5::sig2hex(sig3) << std::endl << std::endl;
+    assert(md5::sig2hex(sig3).compare("e4d909c290d0fb1ca068ffaddf22cbd0") == 0);
+
+    void* sig4 = md5::hash(
```
Comment (Panquesito7) [inline hashing/md5.cpp:327]:
```suggestion
    void* sig4 = hashing::md5::hash(
```

File: hashing/md5.cpp [hunk 27]
```diff
@@ -0,0 +1,352 @@
...
+        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789");
+    std::cout
+        << "Hashing "
+           "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
+        << std::endl;
+    std::cout << md5::sig2hex(sig4) << std::endl << std::endl;
```
Comment (Panquesito7) [inline hashing/md5.cpp:333]:
```suggestion
    std::cout << hashing::md5::sig2hex(sig4) << std::endl << std::endl;
```

File: hashing/md5.cpp [hunk 28]
```diff
@@ -0,0 +1,352 @@
...
+    assert(md5::sig2hex(sig4).compare("d174ab98d277d9f5a5611c2c9f419d9f") == 0);
```
Comment (Panquesito7) [inline hashing/md5.cpp:334]:
```suggestion
    assert(hashing::md5::sig2hex(sig4).compare("d174ab98d277d9f5a5611c2c9f419d9f") == 0);
```

File: hashing/md5.cpp [hunk 29]
```diff
@@ -0,0 +1,352 @@
...
+        void* sig = md5::hash(input);
+        std::cout << "Hash is: " << md5::sig2hex(sig) << std::endl;
+    }
+}
+
+int main() {
+    test();
+    interactive();
+    return 0;
+}
```
Comment (Panquesito7) [inline hashing/md5.cpp:352]:
```suggestion
}

```

File: hashing/md5.cpp [hunk 30]
```diff
@@ -0,0 +1,352 @@
...
+        std::string input;
+        std::cout << "Enter a message to be hashed (only one line): "
+                  << std::endl;
+        std::getline(std::cin, input);
+        void* sig = md5::hash(input);
+        std::cout << "Hash is: " << md5::sig2hex(sig) << std::endl;
+    }
+}
+
+int main() {
```
Comment (Panquesito7) [inline hashing/md5.cpp:348]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
int main() {
```

File: hashing/md5.cpp [hunk 31]
```diff
@@ -0,0 +1,352 @@
...
+    test();
```
Comment (Panquesito7) [inline hashing/md5.cpp:349]:
```suggestion
    test();  // run self-test implementations
```

File: hashing/md5.cpp [hunk 32]
```diff
@@ -0,0 +1,352 @@
...
+        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789");
+    std::cout
+        << "Hashing "
+           "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
+        << std::endl;
+    std::cout << md5::sig2hex(sig4) << std::endl << std::endl;
+    assert(md5::sig2hex(sig4).compare("d174ab98d277d9f5a5611c2c9f419d9f") == 0);
+}
+
+void interactive() {
```
Comment (Panquesito7) [inline hashing/md5.cpp:337]: Missing documentation.

File: hashing/md5.cpp [hunk 33]
```diff
@@ -0,0 +1,352 @@
...
+    while (true) {
+        std::string input;
+        std::cout << "Enter a message to be hashed (only one line): "
+                  << std::endl;
+        std::getline(std::cin, input);
+        void* sig = md5::hash(input);
```
Comment (Panquesito7) [inline hashing/md5.cpp:343]:
```suggestion
        void* sig = hashing::md5::hash(input);
```

File: hashing/md5.cpp [hunk 34]
```diff
@@ -0,0 +1,352 @@
...
+        std::cout << "Hash is: " << md5::sig2hex(sig) << std::endl;
```
Comment (Panquesito7) [inline hashing/md5.cpp:344]:
```suggestion
        std::cout << "Hash is: " << hashing::md5::sig2hex(sig) << std::endl;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 😄👍

**Comment by tGautot** [comment]
Thanks for the feedback!

I have made the modifications you suggested, I hope I was thorough enough.

If there is anything amiss I would be glad to change it!

File: sorting/selection_sort_recursive.cpp [hunk 35]
```diff
@@ -52,7 +52,8 @@ namespace selection_sort_recursive {
  * @returns index of the minimum element
  */
 template <typename T>
-uint64_t findMinIndex(const std::vector<T> &in_arr, uint64_t current_position = 0) {
+uint64_t findMinIndex(const std::vector<T> &in_arr,
+                      uint64_t current_position = 0) {
```
Comment (ayaankhan98) [inline sorting/selection_sort_recursive.cpp:56]: This file is not related to this PR.

**Comment by tGautot** [comment]
I completely removed the problematic commit from the branch to keep everything clean.

File: hashing/md5.cpp [hunk 36]
```diff
@@ -0,0 +1,370 @@
...
+ * some implmenetations can work on bytestream, messages of unknown length.
+ */
+
+#include <algorithm>  // Used for std::copy
+#include <array>      // To avoid c-style array (thanks clang-tidy)
+#include <cassert>
+#include <cstring>  // Used for std::memcopy
+#include <iostream>
+#include <string>
+#include <vector>  // To avoid c-style array of runtime size (thanks clang-tidy)
```
Comment (Panquesito7) [inline hashing/md5.cpp:49]: Please add a one-line description of what the library/header is for (**see the example below**).

```c++
#include <iostream>    /// for IO operations
#include <cassert>    /// for assert
```

File: hashing/md5.cpp [hunk 37]
```diff
@@ -0,0 +1,370 @@
...
+    return hash_bs(&message[0], message.size());
+}
+}  // namespace md5
+}  // namespace hashing
+
+/**
+ * @brief Self-test implementations of well-known MD5 hashes
+ * @returns void
+ */
+void test() {
```
Comment (Panquesito7) [inline hashing/md5.cpp:311]:
```suggestion
static void test() {
```

File: hashing/md5.cpp [hunk 38]
```diff
@@ -0,0 +1,370 @@
...
+        std::cout << "Enter a message to be hashed (only one line): "
+                  << std::endl;
+        std::getline(std::cin, input);
+        void* sig = hashing::md5::hash(input);
+        std::cout << "Hash is: " << hashing::md5::sig2hex(sig) << std::endl;
+    }
+}
+
+/**
+ * @brief Main function, calls the tests and then invokes interactive mode
```
Comment (Panquesito7) [inline hashing/md5.cpp:363]:
```suggestion
 * @brief Main function
 * Calls the tests and then invokes interactive mode
```

File: sorting/selection_sort_recursive.cpp (hunk 35 above)
Comment (Panquesito7) [reply sorting/selection_sort_recursive.cpp:56]: This is modified by the CI automatically; we can't do anything about it. 🙂

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 😄👍

**Comment by tGautot** [comment]
I added the modifications you suggested but it does seem like the CI really wants to change `sorting/selection_sort_recursive.cpp`. If you want me to, I could try re-deleting it.

File: hashing/md5.cpp [hunk 39]
```diff
@@ -0,0 +1,374 @@
...
+#include <array>      // To avoid c-style array (thanks clang-tidy)
+#include <cassert>    // Used for assert
+#include <cstring>    // Used for std::memcopy
+#include <iostream>   // Used for IO operations
+#include <string>     // Used for strings
+#include <vector>  // To avoid c-style array of runtime size (thanks clang-tidy)
+
+/**
+ * @namespace hashing
+ *
```
Comment (Panquesito7) [inline hashing/md5.cpp:53]:
```suggestion
```

File: hashing/md5.cpp [hunk 40]
```diff
@@ -0,0 +1,374 @@
...
+ * @brief Hashing algorithms
+ */
+namespace hashing {
+/**
+ * @namespace MD5
+ *
```
Comment (Panquesito7) [inline hashing/md5.cpp:59]:
```suggestion
```

File: hashing/md5.cpp [hunk 41]
```diff
@@ -0,0 +1,374 @@
...
+
+    void* sig4 = hashing::md5::hash(
+        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789");
+    std::cout
+        << "Hashing "
+           "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
+        << std::endl;
+    std::cout << hashing::md5::sig2hex(sig4) << std::endl << std::endl;
+    assert(hashing::md5::sig2hex(sig4).compare(
+               "d174ab98d277d9f5a5611c2c9f419d9f") == 0);
```
Comment (Panquesito7) [inline hashing/md5.cpp:343]: Add some documentation of what this code does here. 🙂

File: hashing/md5.cpp [hunk 42]
```diff
@@ -0,0 +1,374 @@
...
+        void* sig = hashing::md5::hash(input);
+        std::cout << "Hash is: " << hashing::md5::sig2hex(sig) << std::endl;
+    }
+}
+
+/**
+ * @brief Main function
+ * Calls the tests and then invokes interactive mode
+ * @returns 0 on exit
+ */
```
Comment (Panquesito7) [inline hashing/md5.cpp:366]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

File: sorting/selection_sort_recursive.cpp (hunk 35 above)
Comment (ayaankhan98) [reply sorting/selection_sort_recursive.cpp:56]: This should not happen, When this file was merged why did CI not reformatted this at that time.

Comment (Panquesito7) [reply sorting/selection_sort_recursive.cpp:56]: > This should not happen, When this file was merged why did CI not reformatted this at that time.

That's due to the branch protection.

File: hashing/md5.cpp [hunk 43]
```diff
@@ -0,0 +1,379 @@
...
+ * some implmenetations can work on bytestream, messages of unknown length.
+ */
+
+#include <algorithm>  // Used for std::copy
+#include <array>      // To avoid c-style array (thanks clang-tidy)
+#include <cassert>    // Used for assert
+#include <cstring>    // Used for std::memcopy
+#include <iostream>   // Used for IO operations
+#include <string>     // Used for strings
+#include <vector>  // To avoid c-style array of runtime size (thanks clang-tidy)
```
Comment (Panquesito7) [inline hashing/md5.cpp:47]:
```suggestion
#include <algorithm>  /// Used for std::copy
#include <array>      /// for std::array
#include <cassert>    /// for std::assert
#include <cstring>    /// for std::memcopy
#include <iostream>   /// for IO operations
#include <string>     /// used for strings
#include <vector>     /// for std::vector
```

File: hashing/md5.cpp [hunk 44]
```diff
@@ -0,0 +1,379 @@
...
+/**
+ * @brief Rotates the bits of a 32-bit unsigned integer
+ * @param n Integer to rotate
+ * @param rotate How many bits for the rotation
+ * @return uint32_t The rotated integer
+ */
+uint32_t leftRotate32bits(uint32_t n, std::size_t rotate) {
+    return (n << rotate) | (n >> (32 - rotate));
+}
+
```
Comment (Panquesito7) [inline hashing/md5.cpp:69]:
```suggestion
```

File: hashing/md5.cpp [hunk 45]
```diff
@@ -0,0 +1,379 @@
...
+               "d41d8cd98f00b204e9800998ecf8427e") == 0);
+
+    // Hashes "The quick brown fox jumps over the lazy dog" and stores signature
+    void* sig2 =
+        hashing::md5::hash("The quick brown fox jumps over the lazy dog");
+    std::cout << "Hashing The quick brown fox jumps over the lazy dog"
+              << std::endl;
+    // Prints signature hex representation
+    std::cout << hashing::md5::sig2hex(sig2) << std::endl << std::endl;
+    // Test with cassert wether sig is correct from expected value
```
Comment (Panquesito7) [inline hashing/md5.cpp:320]:
```suggestion
    // Test with cassert whether sig is correct from the expected value
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

File: hashing/md5.cpp [hunk 46]
```diff
@@ -0,0 +1,373 @@
...
+    assert(hashing::md5::sig2hex(sig4).compare(
+               "d174ab98d277d9f5a5611c2c9f419d9f") == 0);
+}
+
+/**
+ * @brief Puts user in a loop where inputs can be given and MD5 hash will be
+ * computed and printed
+ * @returns void
+ */
+static void interactive() {
```
Comment (Panquesito7) [inline hashing/md5.cpp:351]: I like this! It works very well, but I would recommend adding an option for the user to choose if they want to stay inside the interactive mode. I was trying to find a way to quit but then realized it was an infinite loop. 😅

**Comment by tGautot** [comment]
I tried a few things to make interactive mode more clear and enjoyable but all of them would take the focus of the code away from MD5. This is pretty simple but i feel like it does the job, what do you think, maybe you guys have a better idea.

**Comment by Panquesito7** [comment]
> I tried a few things to make interactive mode more clear and enjoyable but all of them would take the focus of the code away from MD5. This is pretty simple but i feel like it does the job, what do you think, maybe you guys have a better idea.

What if we make it like this:

- Write a number.
- Ask the user: want to add another number (y/n)?
- Repeat.

What do you think? 🙂

**Review by Panquesito7 — APPROVED** [review]
Awesome work! 🚀 💯 It's been amazing for your first contribution here! Thanks for contributing! We hope you keep contributing! 😄👍🎉

**Review by amino19 — APPROVED** [review]

**Pull request merged**